docs/topics/numbers.md

```diff
@@ -0,0 +1,289 @@
+
+Kotlin also supports a conventional notation for floating-point numbers:
+
+* Doubles by default: `123.5`, `123.5e10`
+* Floats are tagged by `f` or `F`: `123.5f`
+
+You can use underscores to make number constants more readable:
+
+```kotlin
+val oneMillion = 1_000_000
+val creditCardNumber = 1234_5678_9012_3456L
+val socialSecurityNumber = 999_99_9999L
+val hexBytes = 0xFF_EC_DE_5E
+val bytes = 0b11010010_01101001_10010100_10010010
+```
+
+## Numbers representation on the JVM
+
+On the JVM platform, numbers are stored as primitive types: `int`, `double`, and so on.
+Exceptions are cases when you create a nullable number reference such as `Int?` or use generics.
+In these cases numbers are boxed in Java classes `Integer`, `Double`, and so on.
+
+Note that nullable references to the same number can refer to different objects:
+
+```kotlin
+fun main() {
+//sampleStart
+    val a: Int = 100
+    val boxedA: Int? = a
+    val anotherBoxedA: Int? = a
+    
+    val b: Int = 10000
+    val boxedB: Int? = b
+    val anotherBoxedB: Int? = b
+    
+    println(boxedA === anotherBoxedA) // true
+    println(boxedB === anotherBoxedB) // false
+//sampleEnd
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
+
+All nullable references to `a` are actually the same object because of the memory optimization that JVM applies to `Integer`s
+between `-128` and `127`. It doesn't apply to the `b` references, so they are different objects.
+
+On the other hand, they are still equal:
+
+```kotlin
+fun main() {
+//sampleStart
+    val b: Int = 10000
+    println(b == b) // Prints 'true'
+    val boxedB: Int? = b
+    val anotherBoxedB: Int? = b
+    println(boxedB == anotherBoxedB) // Prints 'true'
+//sampleEnd
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
+
+## Explicit conversions
+
+Due to different representations, smaller types *are not subtypes* of bigger ones.
+If they were, we would have troubles of the following sort:
+
+```kotlin
+// Hypothetical code, does not actually compile:
+val a: Int? = 1 // A boxed Int (java.lang.Integer)
+val b: Long? = a // implicit conversion yields a boxed Long (java.lang.Long)
+print(b == a) // Surprise! This prints "false" as Long's equals() checks whether the other is Long as well
+```
+
+So equality would have been lost silently, not to mention identity.
+
+As a consequence, smaller types _are NOT implicitly converted_ to bigger types.
+This means that assigning a value of type `Byte` to an `Int` variable requires an explicit conversion.
+
+```kotlin
+fun main() {
+//sampleStart
+    val b: Byte = 1 // OK, literals are checked statically
+    // val i: Int = b // ERROR
+    val i1: Int = b.toInt()
+//sampleEnd
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
+
+All number types support conversions to other types:
+
+* `toByte(): Byte`
+* `toShort(): Short`
+* `toInt(): Int`
+* `toLong(): Long`
+* `toFloat(): Float`
+* `toDouble(): Double`
+* `toChar(): Char`
```

Comment:
`toChar` is deprecated, so it makes sense to omit this one.

See https://youtrack.jetbrains.com/issue/KT-50859

Reply:
Thank you, I removed the item. The remaining update requested in the ticket will be in a separate PR.